Windows Process.Name() getFromSnapProcess() performance

Currently `Process.Name()` on Windows calls the `CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, pid)` syscall to get it's ppid and name:

https://github.com/shirou/gopsutil/blob/400a453e67c4f3b5a3a446e788285cedebaf04f4/process/process_windows.go#L321-L322

https://github.com/shirou/gopsutil/blob/400a453e67c4f3b5a3a446e788285cedebaf04f4/process/process_windows.go#L830-L831

It seems `CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, pid)`  will return all processes in the system despite passing the ProcessID parameter.

Although returning a snapshot of  all processes is quite fast most of the time (<10ms),  it can sometimes be slow (>100ms, sometimes even >1000ms).

Can we change the flags like below according to the documentation:
(not so sure about the exact flags, especially `TH32CS_SNAPMODULE32`)
```
func getFromSnapProcess(pid int32) (int32, int32, string, error) {
	// https://learn.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot
	// To enumerate the heap or module states for all processes, specify TH32CS_SNAPALL and set th32ProcessID to zero.
	// Then, for each additional process in the snapshot, call CreateToolhelp32Snapshot again,
	// specifying its process identifier and the TH32CS_SNAPHEAPLIST or TH32_SNAPMODULE value.
	snap, err := windows.CreateToolhelp32Snapshot(windows.TH32CS_SNAPHEAPLIST|windows.TH32CS_SNAPMODULE|windows.TH32CS_SNAPMODULE32, uint32(pid))
	if err != nil {
		return 0, 0, "", err
	}
	defer windows.CloseHandle(snap)
	var pe32 windows.ProcessEntry32
	pe32.Size = uint32(unsafe.Sizeof(pe32))
	if err = windows.Process32First(snap, &pe32); err != nil && err != syscall.ERROR_NO_MORE_FILES {
		return 0, 0, "", err
	}
	for {
		if pe32.ProcessID == uint32(pid) {
			szexe := windows.UTF16ToString(pe32.ExeFile[:])
			return int32(pe32.ParentProcessID), int32(pe32.Threads), szexe, nil
		}
		if err = windows.Process32Next(snap, &pe32); err != nil {
			break
		}
	}
	return 0, 0, "", fmt.Errorf("couldn't find pid: %d", pid)
}
```
or switch to `NtQuerySystemInformation`?

## refs
https://learn.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot:

- TH32CS_SNAPPROCESS 0x00000002 | Includes all processes in the system in the snapshot.

> To enumerate the heap or module states for all processes, specify TH32CS_SNAPALL and set th32ProcessID to zero. 
> Then, for each additional process in the snapshot, call CreateToolhelp32Snapshot again, specifying its process identifier and the **TH32CS_SNAPHEAPLIST** or **TH32_SNAPMODULE** value.

## see also
- https://github.com/Maximus5/ConEmu/issues/885
- https://sudonull.com/post/61039-Speeding-up-the-transfer-of-processes-and-threads-in-Windows-Infopulse-Ukraine-Blog
- https://learn.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntquerysysteminformation

`NtQuerySystemInformation` is really difficult to use, just check #862 (all the cases checking if the calling process and the inspected process are 32 or 64 bit). I won't go out of my way to use it if there's another API with decent performances.

Changing the flag used in `CreateToolhelp32Snapshot` to `windows.TH32CS_SNAPHEAPLIST|windows.TH32CS_SNAPMODULE|windows.TH32CS_SNAPMODULE32` doesn't work unfortunately:

```
go test ./process
--- FAIL: Test_Process_Ppid (0.01s)
    process_test.go:161: getting ppid error Il n’y a plus de fichier.
    process_test.go:164: return value is 0 0
    process_test.go:168: return value is 0, expected 6956
--- FAIL: Test_Process_NumThread (0.00s)
    process_test.go:259: getting NumThread error Il n’y a plus de fichier.
--- FAIL: Test_Process_Threads (0.00s)
    process_test.go:272: getting NumThread error Il n’y a plus de fichier.
    process_test.go:28: not implemented
--- FAIL: Test_Process_Name (0.00s)
    process_test.go:294: getting name error could not get Name: Il n’y a plus de fichier.
    process_test.go:297: invalid Exe
--- FAIL: Test_Process_Long_Name_With_Spaces (1.28s)
    process_test.go:338: getting name error could not get Name: Il n’y a plus de fichier.
--- FAIL: Test_Process_Long_Name (1.35s)
    process_test.go:384: getting name error could not get Name: Il n’y a plus de fichier.
--- FAIL: Test_Parent (0.00s)
    process_test.go:476: error Il n’y a plus de fichier.
FAIL
FAIL    github.com/shirou/gopsutil/v3/process   6.129s
FAIL
```

[In fact, psutil doesn't use `CreateToolhelp32Snapshot` on windows](https://github.com/giampaolo/psutil/blob/8e4099d9f063ceb4ee3da5845562c5b934f83544/psutil/_pswindows.py#L749-L759) to get a process name, but takes the basename of the process executable path. Also, [it is cached after first call](https://github.com/giampaolo/psutil/blob/8e4099d9f063ceb4ee3da5845562c5b934f83544/psutil/__init__.py#L619-L620).

--- a/process/process_test.go
+++ b/process/process_test.go
@@ -294,7 +294,7 @@ func Test_Process_Name(t *testing.T) {
 		t.Errorf("getting name error %v", err)
 	}
 	if !strings.Contains(n, "process.test") {
-		t.Errorf("invalid Exe %s", n)
+		t.Errorf("invalid Name %s", n)
 	}
 }
 

--- a/process/process_windows.go
+++ b/process/process_windows.go
@@ -10,6 +10,7 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"path/filepath"
 	"reflect"
 	"strings"
 	"syscall"
@@ -319,18 +320,19 @@ func (p *Process) PpidWithContext(ctx context.Context) (int32, error) {
 }
 
 func (p *Process) NameWithContext(ctx context.Context) (string, error) {
-	ppid, _, name, err := getFromSnapProcess(p.Pid)
+	if p.Pid == 0 {
+		return "System Idle Process", nil
+	}
+	if p.Pid == 4 {
+		return "System", nil
+	}
+
+	exe, err := p.ExeWithContext(ctx)
 	if err != nil {
 		return "", fmt.Errorf("could not get Name: %s", err)
 	}
 
-	// if no errors and not cached already, cache ppid
-	p.parent = ppid
-	if 0 == p.getPpid() {
-		p.setPpid(ppid)
-	}
-
-	return name, nil
+	return filepath.Base(exe), nil
 }
 
 func (p *Process) TgidWithContext(ctx context.Context) (int32, error) {

--- a/process/process_windows.go
+++ b/process/process_windows.go
@@ -987,15 +987,9 @@ func is32BitProcess(h windows.Handle) bool {
 
 	var procIs32Bits bool
 	switch processorArchitecture {
-	case PROCESSOR_ARCHITECTURE_INTEL:
-		fallthrough
-	case PROCESSOR_ARCHITECTURE_ARM:
+	case PROCESSOR_ARCHITECTURE_INTEL, PROCESSOR_ARCHITECTURE_ARM:
 		procIs32Bits = true
-	case PROCESSOR_ARCHITECTURE_ARM64:
-		fallthrough
-	case PROCESSOR_ARCHITECTURE_IA64:
-		fallthrough
-	case PROCESSOR_ARCHITECTURE_AMD64:
+	case PROCESSOR_ARCHITECTURE_ARM64, PROCESSOR_ARCHITECTURE_IA64, PROCESSOR_ARCHITECTURE_AMD64:
 		var wow64 uint
 
 		ret, _, _ := common.ProcNtQueryInformationProcess.Call(

